velox/functions/lib/Re2Functions.h

```diff
@@ -255,55 +272,90 @@ template <
 struct Re2RegexpReplace {
   VELOX_DEFINE_FUNCTION_TYPES(T);
 
-  std::string processedReplacement_;
-  std::string result_;
-  std::optional<RE2> re_;
-
   FOLLY_ALWAYS_INLINE void initialize(
       const std::vector<TypePtr>& /*inputTypes*/,
       const core::QueryConfig& config,
       const arg_type<Varchar>* /*string*/,
       const arg_type<Varchar>* pattern,
       const arg_type<Varchar>* replacement) {
-    VELOX_USER_CHECK(
-        pattern != nullptr, "Pattern of regexp_replace must be constant.");
-    VELOX_USER_CHECK(
-        replacement != nullptr,
-        "Replacement sequence of regexp_replace must be constant.");
-
-    auto processedPattern = prepareRegexpPattern(*pattern);
-
-    re_.emplace(processedPattern, RE2::Quiet);
-    if (UNLIKELY(!re_->ok())) {
-      VELOX_USER_FAIL(
-          "Invalid regular expression {}: {}.", processedPattern, re_->error());
+    if (pattern != nullptr) {
+      const auto processedPattern = prepareRegexpPattern(*pattern);
+      re_.emplace(processedPattern, RE2::Quiet);
+      VELOX_USER_CHECK(
+          re_->ok(),
+          "Invalid regular expression {}: {}.",
+          processedPattern,
+          re_->error());
     }
 
-    processedReplacement_ = prepareRegexpReplacement(*re_, *replacement);
+    if (replacement != nullptr) {
+      // Constant 'replacement' with non-constant 'pattern' needs to be
+      // processed separately for each row.
+      if (pattern != nullptr) {
+        ensureProcessedReplacement(re_.value(), *replacement);
```

Comment:
nit: I'd prefer `processedReplacement_ = prepareRegexpReplacement(*re_, *replacement);` here, but the current code is OK too.

Reply:
That would require copying replacement even if it is constant and not changing.